File: Graph/GFG/Medium/DijkstraPath/DijkstraPath.cpp
// https://www.geeksforgeeks.org/problems/shortest-path-in-weighted-undirected-graph/1
#include<bits/stdc++.h>
using namespace std;
vector<int> shortestPath(int n, int m, vector<vector<int>> &edges)
{
    // Code here
    set<pair<int, int>> st;
    vector<vector<pair<int, int>>> adj(n + 1);
    for (vector<int> &edge : edges)
    {
        adj[edge[0]].emplace_back(make_pair(edge[1], edge[2]));
        adj[edge[1]].emplace_back(make_pair(edge[0], edge[2]));
    }
    vector<pair<int, int>> minDist(n + 1, make_pair(-1, INT_MAX));
    // previous node along with minimum distance
    minDist[1] = make_pair(-1, 0);
    st.emplace(make_pair(0, 1)); // distance and node

    while (!st.empty())
    {
        pair<int, int> curr = *(st.begin());
        int node = curr.second, dist = curr.first;
        st.erase(curr);

        for (pair<int, int> &neighbor : adj[node])
        {
            int newDist = neighbor.second + dist;

            if (minDist[neighbor.first].second > newDist)
            {
                if (minDist[neighbor.first].second != INT_MAX)
                {
                    st.erase(make_pair(minDist[neighbor.first].second, neighbor.first));
                }
                minDist[neighbor.first] = make_pair(node, newDist);
                st.emplace(make_pair(newDist, neighbor.first));
            }
        }
    }
    if (minDist[n].second == INT_MAX)
        return {-1};
    vector<int> path = {minDist[n].second};
    int node = n;
    while (node != -1)
    {
        path.emplace_back(node);
        node = minDist[node].first;
    }
    if (path.back() != 1)
        return {-1};
    int l = 1, r = path.size() - 1;
    while (l < r)
    {
        swap(path[l], path[r]);
        l++;
        r--;
    }
    return path;
}
int main(void) {
    
    return 0;
}

File: Graph/GFG/Medium/DijkstraPath/DijkstraPath.java
// https://www.geeksforgeeks.org/problems/shortest-path-in-weighted-undirected-graph/1

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.PriorityQueue;

class Pair {

    int first, second;

    Pair(int first, int second) {
        this.first = first;
        this.second = second;
    }
}
public class DijkstraPath {
    public List<Integer> shortestPath(int n, int m, int edges[][]) {
        //  Code Here.
        ArrayList<ArrayList<Pair>> adj = new ArrayList<>();
        
        for(int i=0; i<=n; i++) {
            adj.add(new ArrayList<>());
        }
        
        for(int[] edge: edges) {
            adj.get(edge[0]).add(new Pair(edge[1], edge[2])); // node and weight
            adj.get(edge[1]).add(new Pair(edge[0], edge[2]));
        }
        
        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> {
           return a.first - b.first; // first is distance, second is node
        });
        
        int minDist[][] = new int[n + 1][2]; // first is previous node and second is minDist
        
        for(int i=1; i<=n; i++) {
            minDist[i][1] = Integer.MAX_VALUE;
            minDist[i][0] = -1;
        }
        
        minDist[1][1] = 0;
        
        pq.offer(new Pair(0, 1));
        
        while(!pq.isEmpty()) {
            Pair curr = pq.poll();
            int node = curr.second, dist = curr.first;
            
            pq.remove(curr);
            
            for(Pair neighbor: adj.get(node)) {
                int newDist = neighbor.second + dist;
                if(minDist[neighbor.first][1] > newDist) {
                    minDist[neighbor.first][0] = node;
                    minDist[neighbor.first][1] = newDist;
                    pq.offer(new Pair(newDist, neighbor.first));
                }
            }
        }
        
        if(minDist[n][1] == Integer.MAX_VALUE) {
            List<Integer> ans = new ArrayList<>();
            ans.add(-1);
            return ans;
        }
        
        List<Integer> path = new ArrayList<>();
        path.add(minDist[n][1]);
        int node = n;
        while(node != -1) {
            path.add(node);
            node = minDist[node][0];
        }
        int l = 1, r = path.size() - 1;
        
        while(l < r) {
            Collections.swap(path, l, r);
            l++;
            r--;
        }
        return path;
    }
}
